dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/KafkaInput.java

```diff
@@ -0,0 +1,266 @@
+    private boolean subscribed = false;
+    private volatile boolean running = true;
+    private final ConcurrentLinkedQueue<KafkaConsumerAction> tasks;
+    private final AckTracker ackTracker;
+
+    /**
+     * Lock to synchronize access to the kafkaConsumer instance
+     */
+    private final ReentrantLock lock = new ReentrantLock();
+
+    public KafkaInput(KafkaAdapterFactory kafkaAdapterFactory, KafkaConsumer<K, V> kafkaConsumer, ManagedScheduledExecutorService executor, Collection<String> topics,
+                      int unackedThreshold) {
+        super();
+        this.kafkaConsumer = kafkaConsumer;
+        this.executor = executor;
+        this.topics = topics;
+        this.ackTracker = new AckTracker(kafkaAdapterFactory, executor, unackedThreshold);
+        this.tasks = new ConcurrentLinkedQueue<>();
+        this.ackTracker.setCommitAction(this::commitOffsets);
+    }
+
+    public PublisherBuilder<Message<V>> getPublisher() {
+        if (this.publisher == null) {
+            this.publisher = createPublisher();
+        }
+
+        return this.publisher;
+    }
+
+    private PublisherBuilder<Message<V>> createPublisher() {
+        PublisherBuilder<Message<V>> kafkaStream = ReactiveStreams.generate(() -> 0).flatMapCompletionStage(x -> this.ackTracker.waitForAckThreshold().thenCompose(y -> pollKafkaAsync())).flatMap(Function.identity()).map(this::wrapInMessage).takeWhile((record) -> this.running);
+        return kafkaStream;
+    }
+
+    private CompletionStage<Void> commitOffsets(TopicPartition partition, OffsetAndMetadata offset) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        try {
+            Map<TopicPartition, OffsetAndMetadata> offsets = Collections.singletonMap(partition, offset);
+            runAction((c) -> {
+                c.commitAsync(offsets, (o, e) -> {
+                    if (e != null) {
+                        Tr.warning(tc, "Failed to commit read offsets to kafka broker: {}", e);
+                        result.completeExceptionally(e);
+                    } else {
+                        result.complete(null);
+                    }
+                });
+            });
+            return result;
+        } catch (Throwable t) {
+            Tr.warning(tc, "Failed to commit read offsets to kafka broker: {}", t);
+            result.completeExceptionally(t);
+            return result;
+        }
+    }
+
+    private static <T> Void logPollFailure(T result, Throwable t) {
+        if (t != null) {
+            Tr.error(tc, "Failed to poll Kafka: {}", t);
+        }
+        return null;
+    }
+
+    private Message<V> wrapInMessage(ConsumerRecord<K, V> record) {
+        try {
+            return Message.of(record.value(), this.ackTracker.trackRecord(record));
+        } catch (Throwable t) {
+            Tr.error(tc, "Unexpected error: {}", t);
+            throw t;
+        }
+    }
+
+    public void shutdown() {
```

Comment:
This shutdown method is never called. @Azquelt to open a new issue to consider general shutdown/cleanup.

Reply:
#8006